lib/Scheduling/CPSATSchedulers.cpp

```diff
@@ -0,0 +1,138 @@
+// via OR-Tools.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Scheduling/Algorithms.h"
+
+#include "mlir/IR/Operation.h"
+
+#include "ortools/sat/cp_model.h"
+
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Format.h"
+
+#define DEBUG_TYPE "cpsat-schedulers"
+
+using namespace circt;
+using namespace circt::scheduling;
+using namespace operations_research;
+using namespace operations_research::sat;
+
+using llvm::dbgs;
+using llvm::format;
+
+LogicalResult scheduling::scheduleCPSAT(SharedOperatorsProblem &prob,
+                                        Operation *lastOp) {
+  Operation *containingOp = prob.getContainingOp();
+  if (!prob.hasOperation(lastOp))
+    return containingOp->emitError("problem does not include last operation");
+
+  CpModelBuilder cpModel;
+  auto &tasks = prob.getOperations();
+  auto &resources = prob.getOperatorTypes();
+
+  DenseMap<Operation *, IntVar> taskStarts;
+  DenseMap<Operation *, IntVar> taskEnds;
+  DenseMap<Problem::OperatorType, llvm::SmallVector<IntervalVar, 4>>
+      resourcesToTaskIntervals;
+
+  // first get horizon
+  unsigned horizon = 0;
+  for (auto *task : tasks) {
+    unsigned duration = *prob.getLatency(*prob.getLinkedOperatorType(task));
+    horizon += duration;
+  }
+
+  // build vars
+  for (auto item : llvm::enumerate(tasks)) {
+    auto i = item.index();
+    auto *task = item.value();
+    IntVar startVar = cpModel.NewIntVar(Domain(0, horizon))
+                          .WithName((Twine("start_of_task_") + Twine(i)).str());
+    IntVar endVar = cpModel.NewIntVar(Domain(0, horizon))
+                        .WithName((Twine("end_of_task_") + Twine(i)).str());
+    taskStarts[task] = startVar;
+    taskEnds[task] = endVar;
+    auto resource = prob.getLinkedOperatorType(task);
+    if (resource.hasValue()) {
+      unsigned duration = *prob.getLatency(*resource);
+      IntervalVar taskInterval =
+          cpModel.NewIntervalVar(startVar, duration, endVar)
+              .WithName((Twine("task_interval_") + Twine(i)).str());
+
+      if (!resourcesToTaskIntervals.count(resource.value()))
+        resourcesToTaskIntervals[resource.value()] = {};
+      resourcesToTaskIntervals[resource.value()].emplace_back(taskInterval);
+    }
+  }
+
+  IntVar makeSpan = cpModel.NewIntVar(Domain(0, horizon)).WithName("makespan");
+  for (Operation *task : tasks) {
+    cpModel.AddLessOrEqual(taskEnds[task], makeSpan);
+    for (auto dep : prob.getDependences(task)) {
+      Operation *src = dep.getSource();
+      Operation *dst = dep.getDestination();
+      if (src == dst)
+        return containingOp->emitError() << "dependence cycle detected";
+      cpModel.AddLessOrEqual(taskEnds[src], taskStarts[dst]);
+    }
+  }
+
+  for (auto resourceToTaskIntervals : resourcesToTaskIntervals) {
+    // https://or.stackexchange.com/a/3363
+    const Problem::OperatorType &resource = resourceToTaskIntervals.getFirst();
```

Comment:
Please check if `const` is needed (here and below), it is at least uncommon in the code base to have local consts (though I'm not sure if it's a hard rule).